tasks/storage-market/market_balance.go

```diff
@@ -0,0 +1,189 @@
+		api:    api,
+		cfg:    &mk12Cfg,
+		miners: mmap,
+		sender: sender,
+	}, nil
+}
+
+func (m *MarketBalanceManager) Do(taskID harmonytask.TaskID, stillOwned func() bool) (done bool, err error) {
+	llog := log.With("Task", "MarketBalanceManager")
+
+	ctx := context.Background()
+
+	threshold := abi.TokenAmount(m.cfg.CollateralAddThreshold)
+	amount := abi.TokenAmount(m.cfg.CollateralAddAmount)
+
+	for module, miners := range m.miners {
+		if module != mk12Str {
+			continue
+		}
+		for _, miner := range miners {
+			miner := miner
+
+			// Check head in loop in case it changes and wallet balance changes
+			head, err := m.api.ChainHead(ctx)
+			if err != nil {
+				return false, xerrors.Errorf("failed to get chain head: %w", err)
+			}
+
+			bal, err := m.api.StateMarketBalance(ctx, miner, head.Key())
+			if err != nil {
+				return false, xerrors.Errorf("failed to get market balance: %w", err)
+			}
+
+			avail := big.Sub(bal.Escrow, bal.Locked)
+
+			if avail.GreaterThan(threshold) {
+				llog.Debugf("Skipping add balance for miner %s, available balance is %s, threshold is %s", miner.String(), avail.String(), threshold.String())
+				continue
+			}
+
+			var sender address.Address
+
+			if m.cfg.DealCollateralWallet != "" {
+				wallet, err := address.NewFromString(m.cfg.DealCollateralWallet)
+				if err != nil {
+					return false, xerrors.Errorf("failed to parse deal collateral wallet: %w", err)
+				}
+
+				w, err := m.api.StateGetActor(ctx, wallet, head.Key())
+				if err != nil {
+					return false, xerrors.Errorf("failed to get wallet actor: %w", err)
+				}
+
+				if w.Balance.LessThan(amount) {
+					return false, xerrors.Errorf("Collateral wallet balance %s is lower than specified amount %s", w.Balance.String(), amount.String())
+				}
+				sender = wallet
+			} else {
+				mi, err := m.api.StateMinerInfo(ctx, miner, head.Key())
+				if err != nil {
+					return false, xerrors.Errorf("failed to get miner info: %w", err)
+				}
+
+				w, err := m.api.StateGetActor(ctx, mi.Worker, head.Key())
+				if err != nil {
+					return false, xerrors.Errorf("failed to get wallet actor: %w", err)
+				}
+
+				if w.Balance.LessThan(amount) {
+					return false, xerrors.Errorf("Worker wallet balance %s is lower than specified amount %s", w.Balance.String(), amount.String())
+				}
+				sender = mi.Worker
+			}
+
+			params, err := actors.SerializeParams(&miner)
+			if err != nil {
+				return false, xerrors.Errorf("failed to serialize miner address: %w", err)
+			}
+
+			maxfee, err := types.ParseFIL("0.05 FIL")
+			if err != nil {
+				return false, xerrors.Errorf("failed to parse max fee: %w", err)
+			}
+
+			msp := &lapi.MessageSendSpec{
+				MaxFee: abi.TokenAmount(maxfee),
+			}
+
+			msg := &types.Message{
+				To:     marketActor.Address,
+				From:   sender,
+				Value:  amount,
+				Method: marketActor.Methods.AddBalance,
+				Params: params,
+			}
+
+			mcid, err := m.sender.Send(ctx, msg, msp, "Add Market Collateral")
```

Comment:
All other places use this-kind-of-case, best to stay consistent, later on we can create a nice webui for viewing message sends by address/type/etc, showing e.g. where gas fees go

```suggestion
			mcid, err := m.sender.Send(ctx, msg, msp, "add-market-collateral")
```